stable_baselines3/common/recurrent/torch_layers.py

```diff
@@ -0,0 +1,205 @@
+import abc
+from typing import Any, Dict, Generic, Optional, Tuple, TypeVar
+
+import gymnasium as gym
+import torch as th
+
+from stable_baselines3.common.preprocessing import get_flattened_obs_dim
+from stable_baselines3.common.pytree_dataclass import TensorTree, tree_flatten, tree_map
+from stable_baselines3.common.recurrent.type_aliases import (
+    GRURecurrentState,
+    LSTMRecurrentState,
+)
+from stable_baselines3.common.torch_layers import (
+    BaseFeaturesExtractor,
+    CombinedExtractor,
+    FlattenExtractor,
+    NatureCNN,
+)
+from stable_baselines3.common.type_aliases import TorchGymObs
+
+RecurrentState = TypeVar("RecurrentState", bound=TensorTree)
+
+RecurrentSubState = TypeVar("RecurrentSubState", bound=TensorTree)
+
+ExtractorInput = TypeVar("ExtractorInput", bound=TorchGymObs)
+
+
+class RecurrentFeaturesExtractor(BaseFeaturesExtractor, abc.ABC, Generic[ExtractorInput, RecurrentState]):
+    @abc.abstractmethod
+    def recurrent_initial_state(
+        self, n_envs: Optional[int] = None, *, device: Optional[th.device | str] = None
+    ) -> RecurrentState:
+        ...
+
+    @abc.abstractmethod
+    def forward(
+        self, observations: ExtractorInput, state: RecurrentState, episode_starts: th.Tensor
+    ) -> Tuple[th.Tensor, RecurrentState]:
+        ...
+
+    @staticmethod
+    def _process_sequence(
+        rnn: th.nn.RNNBase, inputs: th.Tensor, init_state: RecurrentSubState, episode_starts: th.Tensor
+    ) -> Tuple[th.Tensor, RecurrentSubState]:
+        (state_example, *_), _ = tree_flatten(init_state, is_leaf=None)
+        n_layers, batch_sz, *_ = state_example.shape
+
+        # Batch to sequence
+        # (padded batch size, features_dim) -> (n_seq, max length, features_dim) -> (max length, n_seq, features_dim)
```

Comment:
I think this could be more succinctly expressed w/ einops.rearrange. Ex:
```
seq_inputs = einops.rearrange(inputs, "(batch seqlen) ... -> seqlen batch ...", batch=batch_sz)
```

But I don't have a terribly strong opinion, particularly if you just want to stay away from einops as an unnecessary dependency.

Reply:
I'm going to make this less complicated very soon, so I would like to leave it like it is. But I agree that `einops` would make it easier to read.

Reply:
I think we should wait until the next time `rearrange` is obviously useful here.

Reply:
SGTM, pulling in a new dependency for one line seems like overkill